AppManager/app_manager_server.py

```diff
@@ -767,6 +818,204 @@ def create_java_start_cmd(app_name, port, load_balancer_host, max_heap,
   return ' '.join(cmd)
 
 
+def unregister_instance(instance):
+  """ Removes a registration entry for an instance.
+
+  Args:
+    instance: An Instance.
+  """
+  instance_entry = ':'.join([options.private_ip, str(instance.port)])
+  instance_node = '/'.join([VERSION_REGISTRATION_NODE, instance.version_key,
+                            instance_entry])
+
+  try:
+    zk_client.delete(instance_node)
+  except NoNodeError:
+    pass
+
+  running_instances.remove(instance)
+
+
+def register_instance(instance):
+  """ Adds a registration entry for an instance.
+
+  Args:
+    instance: An Instance.
+  """
+  instance_entry = ':'.join([options.private_ip, str(instance.port)])
+  instance_node = '/'.join([VERSION_REGISTRATION_NODE, instance.version_key,
+                            instance_entry])
+
+  try:
+    zk_client.create(instance_node, instance.revision.encode('utf-8'))
+  except NodeExistsError:
+    zk_client.set(instance_node, instance.revision.encode('utf-8'))
+
+  running_instances.add(instance)
+
+
+def declare_instance_nodes(running_instances, zk_client):
+  """ Removes dead ZooKeeper instance entries and adds running ones.
+
+  Args:
+    running_instances: An iterable of Instances.
+    zk_client: A KazooClient.
+  """
+  registered_instances = set()
+  for version_key in zk_client.get_children(VERSION_REGISTRATION_NODE):
+    version_node = '/'.join([VERSION_REGISTRATION_NODE, version_key])
+    for instance_entry in zk_client.get_children(version_node):
+      machine_ip, port = instance_entry.split(':')
+      port = int(port)
```

Comment:
sorry couldn't resist: perhaps move this after the check? No need to initialize it if we are not the right compute node.